vcd/datasource_vcd_vapp_test.go

```diff
@@ -0,0 +1,114 @@
+// +build vapp ALL functional
+
+package vcd
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+)
+
+// getAvailableVapp collects one available Vapp to use in data source tests
+func getAvailableVapp() (*govcd.VApp, error) {
+	// Get the data from configuration file. This client is still inactive at this point
+	vcdClient, err := getTestVCDFromJson(testConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error getting client configuration: %s", err)
+	}
+	err = vcdClient.Authenticate(testConfig.Provider.User, testConfig.Provider.Password, testConfig.Provider.SysOrg)
+	if err != nil {
+		return nil, fmt.Errorf("authentication error: %s", err)
+	}
+	org, err := vcdClient.GetAdminOrgByName(testConfig.VCD.Org)
+	if err != nil {
+		return nil, fmt.Errorf("org not found : %s", err)
+	}
+	vdc, err := org.GetVDCByName(testConfig.VCD.Vdc, false)
+	if err != nil {
+		return nil, fmt.Errorf("vdc not found : %s", err)
+	}
+
+	for _, resourceEntities := range vdc.Vdc.ResourceEntities {
+		for _, resourceReference := range resourceEntities.ResourceEntity {
+			if resourceReference.Type == "application/vnd.vmware.vcloud.vApp+xml" {
+				return vdc.GetVAppByHref(resourceReference.HREF)
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("no vApp found in VDC %s", testConfig.VCD.Vdc)
+}
+
+// TestAccVcdVappDS tests a vApp data source if a vApp is found in the VDC
+func TestAccVcdVappDS(t *testing.T) {
+	// This test requires access to the vCD before filling templates
+	// Thus it won't run in the short test
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+
+	vapp, err := getAvailableVapp()
+	if err != nil {
+		t.Skip("No suitable vApp found for this test")
+		return
+	}
+
+	var params = StringMap{
+		"Org":      testConfig.VCD.Org,
+		"VDC":      testConfig.VCD.Vdc,
+		"VappName": vapp.VApp.Name,
+		"FuncName": "TestVappDS",
+		"Tags":     "vapp",
+	}
+	configText := templateFill(datasourceTestVapp, params)
+	debugPrintf("#[DEBUG] CONFIGURATION: %s", configText)
+
+	statusText, err := vapp.GetStatus()
+	if err != nil {
+		statusText = vAppUnknownStatus
+	}
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: configText,
+				Check: resource.ComposeTestCheckFunc(
+
+					resource.TestCheckOutput("name", vapp.VApp.Name),
+					resource.TestCheckOutput("description", vapp.VApp.Description),
+					resource.TestCheckOutput("href", vapp.VApp.HREF),
+					resource.TestCheckOutput("status_text", statusText),
+					resource.TestCheckResourceAttr("data.vcd_vapp."+vapp.VApp.Name, "status", fmt.Sprintf("%d", vapp.VApp.Status)),
```

Comment:
Could we test for all/more parameters? I'm especially concerned about `guest_properties`.

Reply:
I've tested what is not deprecated.

Regarding `guest_properties`, the way it works now, requires a setup similar to what we do for load balancers. We need to set the guest properties, then add a VM, use a initscript to set a HTTP server, and then get the result of the guest properties from the server. Not a trivial task. This goes beyond the purpose of this PR.


cc: @Didainius

Reply:
We have already, however, a simpler test for `guest properties` in  `resource_vcd_vapp_vm_properties_test.go` and `resource_vcd_vapp_properties_test.go`

Reply:
> Regarding guest_properties, the way it works now, requires a setup similar to what we do for load balancers.

That's not needed here. All I'm asking is to set them and read them out to see that we have set them correctly, similar to the way we're doing for VM in vcd/resource_vcd_vapp_vm_properties_test.go but for vApp.

Reply:
I understood that you wanted to test the true purpose of `guest_properties` in vApp.

If we just need to check whether the properties are set, isn't what `resource_vcd_vapp_properties_test.go` does?